components/layout_thread_2020/lib.rs

```diff
@@ -1261,16 +1262,30 @@ impl FontMetricsProvider for LayoutFontMetricsProvider {
 
             // Only use the x-height of this font if it is non-zero. Some fonts return
             // inaccurate metrics, which shouldn't be used.
-            let x_height = Some(servo_metrics.x_height)
+            let x_height = Some(first_font_metrics.x_height)
                 .filter(|x_height| !x_height.is_zero())
                 .map(CSSPixelLength::from);
 
+            const ZERO_CODEPOINT: char = '0';
+            const IC_CODEPOINT: char = '\u{6C34}';
+            let [zero_advance_measure, ic_width] =
+                [ZERO_CODEPOINT, IC_CODEPOINT].map(|codepoint| {
+                    let font = font_group
+                        .borrow_mut()
+                        .find_by_codepoint(font_context, codepoint)?;
+                    let font = font.borrow();
+                    let glyph_index = font.glyph_index(codepoint)?;
+                    // TODO: Use the vertical advance for vertical writing modes.
+                    let advance = font.glyph_h_advance(glyph_index);
+                    Some(CSSPixelLength::new(advance as f32))
+                });
+
```

Comment:
I think it would be better to move this to the `font.rs` and have this calculated when creating a font versus here, late in the process.

Reply:
For both of these units, any given font in the font group might not have a glyph for the character, and the required advance is that for the first available font which has it. (While it's unlikely for any font to not have a zero glyph, this is much more likely for the CJK water ideograph.) So these advances are a property of the font group, not of a font.

Would it make sense to compute the advances of these two characters for every font, and to resolve them when creating a `FontGroup`?

Reply:
I think the best approach in that case is to make that field in the `FontMetrics` an `Option`. That way we only fetch this glyph once when the font is initialized rather than every time the style system needs it.

Reply:
Here's the relevant code in Gecko for FreeType: [gfxFT2FontBase.cpp](https://searchfox.org/mozilla-central/rev/0916ef0172ce5b2a72749b659da8ad95f637ef42/gfx/thebes/gfxFT2FontBase.cpp#480).